onnx/onnx.proto

```diff
@@ -297,6 +297,68 @@ message TensorProto {
+    QUINT8 = 16;
+    // Quantized tensor with UINT16 elements, scalar UINT16 zero point and
+    // scalar FLOAT scale. Zero point and scale are static scalar quantization
+    // parameters shared by all elements of a tensor. Elements of a quantized
+    // tensor have real-valued representation r[i] = (q[i] - zero_point) * scale
+    // where q[i] is an element of a quantized tensor and r[i] is its
+    // real-valued representation.
+    //
+    // A compliant ONNX implementation can use an equivalent representation of
+    // quantized tensors, e.g. as 17-bit signed integer elements + FLOAT scale,
+    // or as FLOAT16 elements. If a floating-point representation is used,
+    // an ONNX implementation must guarantee that all quantized elements are
+    // within [-zero_point * scale, (65535 - zero_point) * scale] range.
+    QUINT16 = 17;
+    // Quantized tensor with INT8 elements and scalar FLOAT scale. Scale is a
+    // static scalar quantization parameter shared by all elements of a tensor.
+    // Elements of a quantized tensor have real-valued representation
+    // r[i] = q[i] * scale where q[i] is an element of a quantized tensor and
+    // r[i] is its real-valued representation.
+    //
+    // A compliant ONNX implementation can use an equivalent representation of
+    // quantized tensors, e.g. as FLOAT16 elements. If a floating-point
+    // representation is used, an ONNX implementation must guarantee that all
+    // quantized elements are within [-128 * scale, 127 * scale] range.
+    QINT8 = 18;
+    // Quantized tensor with INT16 elements and scalar FLOAT scale. Scale is a
+    // static scalar quantization parameter shared by all elements of a tensor.
+    // Elements of a quantized tensor have real-valued representation
+    // r[i] = q[i] * scale where q[i] is an element of a quantized tensor and
+    // r[i] is its real-valued representation.
+    //
+    // A compliant ONNX implementation can use an equivalent representation of
+    // quantized tensors, e.g. as FLOAT16 elements. If a floating-point
+    // representation is used, an ONNX implementation must guarantee that all
+    // quantized elements are within [-32768 * scale, 32767 * scale] range.
+    QINT16 = 19;
+    // Quantized tensor with INT32 elements and scalar FLOAT scale. Scale is a
+    // static scalar quantization parameter shared by all elements of a tensor.
+    // Elements of a quantized tensor have real-valued representation
+    // r[i] = q[i] * scale where q[i] is an element of a quantized tensor and
+    // r[i] is its real-valued representation.
+    //
+    // A compliant ONNX implementation can use an equivalent representation of
+    // quantized tensors, e.g. as FLOAT16 elements. If a floating-point
+    // representation is used, an ONNX implementation must guarantee that all
+    // quantized elements are within [-2147483648 * scale, 2147483647 * scale]
+    // range.
+    QINT32 = 20;
+
```

Comment:
I'm not seeing any additions to TensorProto regarding scale and zero_point for the sake of storing static tensors in initializers. If one is expected to correlate the TensorProto.name to the Graph::value_info's just to extract the scale and zero_point, that's kinda indirect, and it's weird because both TensorProto.dims and TypeProto.Tensor.shape store shape information. Which wins?

Reply:
The reader of the graph is expected to match `TensorProto.name` to `GraphProto.input.name` to extract `scale` and `zero_point`. I'm open to add `QuantizationParameters` to `TensorProto` as well, but note that `GraphProto.initializer` and `TensorProto` are not always present (e.g. in ONNXIFI weights are commonly passed in memory rather then encoded into protobuf).